javalib/src/main/scala/java/lang/String.scala

```diff
@@ -667,22 +668,598 @@ final class _String()
+     */
+    def afterI(i: Int): scala.Boolean = {
+      val j = skipCharsWithCombiningClassOtherThanNoneOrAboveBackwards(i)
+      j > 0 && charAt(j - 1) == 'I'
+    }
+
+    /* Tests whether we are in an `Before_Dot` context.
+     * From Table 3.17 in the Unicode standard:
+     * - Description: C is followed by combining dot above (U+0307). Any
+     *   sequence of characters with a combining class that is neither 0 nor
+     *   230 may intervene between the current character and the combining dot
+     *   above.
+     * - Regex, after C: ([^\p{ccc=230}\p{ccc=0}])*[\u0307]
+     */
+    def beforeDot(i: Int): scala.Boolean = {
+      val j = skipCharsWithCombiningClassOtherThanNoneOrAboveForwards(i + 1)
+      j != length() && charAt(j) == '\u0307'
+    }
+
+    val preprocessed = replaceCharsAtIndex { i =>
+      (this.charAt(i): @switch) match {
+        case '\u0130'                  => "\u0069"
+        case '\u0307' if afterI(i)     => ""
+        case '\u0049' if !beforeDot(i) => "\u0131"
+        case _                         => null
+      }
+    }
+
+    preprocessed.toLowerCase()
+  }
+
+  @inline def toLowerCase(): _String = {
+    replaceCharsAtIndex { i =>
+      // Test if given character is last cased letter within given word context, that means
+      // given char at index has at least 1 preceding cased letter within word
+      // and is not followed by any cased letter
+      def isFinalCased(idx: Int): Boolean = {
+
+        /* Character is cased when its type matches lowercase, uppercase or title-case type.
+         * Alternatively it can be non standard cased character, eq. modifier or circled letter or roman number
+         */
+        def isCased(c: Char): scala.Boolean = {
+          val charType = Character.getType(c)
+          charType == Character.LOWERCASE_LETTER ||
+          charType == Character.UPPERCASE_LETTER ||
+          charType == Character.TITLECASE_LETTER ||
+          ((c >= 0x02B0) && (c <= 0x02B8)) || // Modifier small letter h..y
+          ((c >= 0x02C0) && (c <= 0x02C1)) || // Modifier letter or reversed letter glottal stop
+          ((c >= 0x02E0) && (c <= 0x02E4)) || // Modifier small or reversed small letter gamma
+          c == '\u0345' ||                    // Combining greek ypogegrammeni
+          c == '\u037A' ||                    // Greek ypogegrammeni
+          ((c >= 0x1D2C) && (c <= 0x1D61)) || // Modifier letter capital A to CHI
+          ((c >= 0x2160) && (c <= 0x217F)) || // Roman capital or small numeral one to one thousand
+          ((c >= 0x24B6) && (c <= 0x24E9))    // Circled latin capital or small letters A to Z
+        }
+
+        @tailrec
+        def followsCased(i: Int): scala.Boolean = {
+          if (i < offset) false // Not found non cased in whole string
+          else {
+            val c = charAt(i)
+            // Outside word boundary and not found cased
+            if (c.isWhitespace) false
```

Comment:
Where does `isWhitespace` come from? I see no reference to whitespace in the Unicode spec for `Final_Sigma`. However I see the concept of "case-ignorable", which I don't see in the code.

We should stick 100% to the Unicode spec of `Final_Sigma`, and comment everything with the references that we're using to make the code directly comparable to the spec.

Reply:
You're right, it was a kind of simplification used to match with JVM, eg. matching only characters within word boundary. I've replaced it with the correct implementation based on Unicode specification. To make it possible I had to add two additional methods to `Characters`  and associated arrays of range indices.

Reply:
> You're right, it was a kind of simplification used to match with JVM, eg. matching only characters within word boundary.

Hum, wait. Do you mean that the JVM's `toLowerCase()` behaves differently than the Unicode spec? Do you have an example input string where it's not the same?

Reply:
Not really, by simplification I've meant simplification on our side, eg. test ```assertEquals("dσς aσς bσc", "DΣΣ AΣΣ BΣC".toLowerCase())`` was created as I've seen that only last sigma in each word should be changed and didn't know or interpretted incorectlly `caseIgnorable` rule from specification. 

In each case I've tested on JVM (for Final_Sigma rule) incompatibilites with Unicode spec comes from implementation of different Unicode version in given JDK. For example there are 181 more "cased" symbols in Unicode 13.0.0 when comparing to 10.0.0 used in JDK 11. I belive it's fixed in newer versions.